File: 智能优化算法及其MATLAB实例/智能优化算法第2版-源程序/chapter3 差分进化算法Differential Evolution/DE31.m
%%%%%%%%%%%%%%%%%差分进化算法求函数极值%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%初始化%%%%%%%%%%%%%%%%%%%%%%%%%
clear all;                            %清除所有变量
close all;                            %清图
clc;                                  %清屏
NP=50;                                %个体数目
D=10;                                 %变量的维数
G=200;                                %最大进化代数
F0=0.4;                               %初始变异算子
CR=0.1;                               %交叉算子
Xs=20;                                %上限
Xx=-20;                               %下限
yz=10^-6;                             %阈值
%%%%%%%%%%%%%%%%%%%%%%%%%赋初值%%%%%%%%%%%%%%%%%%%%%%%%
x=zeros(D,NP);                        %初始种群
v=zeros(D,NP);                        %变异种群
u=zeros(D,NP);                        %选择种群
x=rand(D,NP)*(Xs-Xx)+Xx;              %赋初值
   %%%%%%%%%%%%%%%%%%%%计算目标函数%%%%%%%%%%%%%%%%%%%%
for m=1:NP
    Ob(m)=func1(x(:,m));
end
trace(1)=min(Ob);
%%%%%%%%%%%%%%%%%%%%%%%差分进化循环%%%%%%%%%%%%%%%%%%%%%
for gen=1:G
    %%%%%%%%%%%%%%%%%%%%%%变异操作%%%%%%%%%%%%%%%%%%%%%%
    %%%%%%%%%%%%%%%%%%%自适应变异算子%%%%%%%%%%%%%%%%%%%
    lamda=exp(1-G/(G+1-gen));
    F=F0*2^(lamda);
    %%%%%%%%%%%%%%%%%r1,r2,r3和m互不相同%%%%%%%%%%%%%%%%
    for m=1:NP
        r1=randi([1,NP],1,1);
        while (r1==m)
            r1=randi([1,NP],1,1);
        end
        r2=randi([1,NP],1,1);
        while (r2==m)|(r2==r1)
            r2=randi([1,NP],1,1);
        end
        r3=randi([1,NP],1,1);
        while (r3==m)|(r3==r1)|(r3==r2)
            r3=randi([1,NP],1,1);
        end
        v(:,m)=x(:,r1)+F*(x(:,r2)-x(:,r3));
    end
    %%%%%%%%%%%%%%%%%%%%%%交叉操作%%%%%%%%%%%%%%%%%%%%%%%
    r=randi([1,D],1,1);
    for n=1:D
        cr=rand(1);
        if (cr<=CR)|(n==r)
            u(n,:)=v(n,:);
        else
            u(n,:)=x(n,:);
        end
    end
    %%%%%%%%%%%%%%%%%%%边界条件的处理%%%%%%%%%%%%%%%%%%%%%
    for n=1:D
        for m=1:NP
            if (u(n,m)<Xx)|(u(n,m)>Xs)
                u(n,m)=rand*(Xs-Xx)+Xx;
            end
        end
    end
    %%%%%%%%%%%%%%%%%%%%%%选择操作%%%%%%%%%%%%%%%%%%%%%%%
    for m=1:NP
        Ob1(m)=func1(u(:,m));
    end
    for m=1:NP
        if Ob1(m)<Ob(m)
            x(:,m)=u(:,m);
        end
    end  
    for m=1:NP
        Ob(m)=func1(x(:,m));
    end
    trace(gen+1)=min(Ob);
    if min(Ob(m))<yz
        break
    end
end
[SortOb,Index]=sort(Ob);
x=x(:,Index);
X=x(:,1);                              %最优变量              
Y=min(Ob);                             %最优值  
%%%%%%%%%%%%%%%%%%%%%%%%%画图%%%%%%%%%%%%%%%%%%%%%%%%%%
figure
plot(trace);
xlabel('迭代次数')
ylabel('目标函数值')
title('适应度进化曲线')

File: 智能优化算法及其MATLAB实例/智能优化算法第2版-源程序/chapter3 差分进化算法Differential Evolution/DE31.py
import numpy as np
import matplotlib.pyplot as plt
plt.rcParams["font.sans-serif"]=["SimHei"] #设置字体
plt.rcParams["axes.unicode_minus"]=False #该语句解决图像中的“-”负号的乱码问题
def func1(x):
    # 根据具体的目标函数定义实现函数体
    return np.sum(x**2) # 示例: 目标函数为每个变量的平方和

def de_algorithm():
    # 初始化参数
    NP = 50      # 个体数目
    D = 10       # 变量的维数
    G = 200      # 最大进化代数
    F0 = 0.4     # 初始变异算子
    CR = 0.1     # 交叉算子
    Xs = 20      # 上限
    Xx = -20     # 下限
    yz = 1e-6    # 阈值

    # 赋初值
    x = np.random.uniform(low=Xx, high=Xs, size=(D, NP))

    # 计算目标函数
    Ob = np.zeros(NP)
    for m in range(NP):
        Ob[m] = func1(x[:, m])
    trace = [np.min(Ob)]

    # 差分进化循环
    for gen in range(G):
        # 变异操作
        lamda = np.exp(1 - G / (G + 1 - gen))
        F = F0 * 2**lamda

        v = np.zeros((D, NP))
        for m in range(NP):
            r = np.random.choice(np.arange(NP), size=3, replace=False)
            while r[0] == m or r[1] == m or r[2] == m:
                r = np.random.choice(np.arange(NP), size=3, replace=False)
            v[:, m] = x[:, r[0]] + F * (x[:, r[1]] - x[:, r[2]])

        # 交叉操作
        r = np.random.randint(0, D)
        u = np.where(np.random.rand(D, NP) <= CR, v, x)
        u[r, :] = v[r, :]

        # 处理边界条件
        u = np.clip(u, Xx, Xs)

        # 选择操作
        Ob1 = np.zeros(NP)
        for m in range(NP):
            Ob1[m] = func1(u[:, m])
            if Ob1[m] < Ob[m]:
                x[:, m] = u[:, m]

        Ob = np.copy(Ob1)
        trace.append(np.min(Ob))

        if np.min(Ob) < yz:
            break

    SortOb = np.sort(Ob)
    Index = np.argsort(Ob)
    x = x[:, Index]
    X = x[:, 0]
    Y = np.min(Ob)
    print("X={}".format(X))
    print("Y={}".format(Y))

    # 画图
    plt.plot(trace)
    plt.xlabel('迭代次数')
    plt.ylabel('目标函数值')
    plt.title('适应度进化曲线')
    plt.show()

de_algorithm()
